### PR TITLE
Add transform function

### PR DESCRIPTION
**Pull request by Akado2009**
- introduces new function - transform (actually transforms FT to NFT) to facilitate some concert functionality
- added comments & docs
- also a test for a new function

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -106,11 +130,87 @@ impl SimpleMTKCore for SimpleMTK {
...
+        nfts_metadata: Vec<Vec<Option<TokenMetadata>>>,
+    ) {
+        // prechecks
+        if nft_ids.len() != nfts_metadata.len() {
+            panic!("MTK: number of ids, accounts and metadata pieces MUST be the same.");
+        }
+        let mut nft_count = 0;
+        for (_, ids) in nft_ids.iter().enumerate() {
+            nft_count += ids.len();
+        }
```
Comment (shamilsan) [inline src/lib.rs]: Why to enumerate and to not use an index?
```suggestion
        for ids in &nft_ids {
            nft_count += ids.len();
        }
```

File: src/lib.rs [hunk 2]
```diff
@@ -106,11 +130,87 @@ impl SimpleMTKCore for SimpleMTK {
...
+
+        let sup = self.supply(id);
+        let mut _balance = self
+            .supply
+            .insert(self.token_id, sup.saturating_sub(amount));
+        let mut ids = vec![];
+        // now we need to mint NFTs (same - just use mint_implementation to remove excessive usage)
+        for (idx, metadata) in nfts_metadata.iter().enumerate() {
+            let amount = nft_ids[idx].len();
+            if accounts[idx] == ActorId::zero() {
```
Comment (shamilsan) [inline src/lib.rs]: You can use `Actor::is_zero` function.

File: src/lib.rs [hunk 3]
```diff
@@ -106,11 +130,87 @@ impl SimpleMTKCore for SimpleMTK {
...
+                panic!("MTK: Mint to zero address")
+            }
+            if nft_ids[idx].len() != amount {
```
Comment (shamilsan) [inline src/lib.rs]: I don't understand this part.
```rust
let amount = nft_ids[idx].len();
if nft_ids[idx].len() != amount ...
```

File: tests/utils.rs [hunk 4]
```diff
@@ -239,6 +239,43 @@ pub fn transfer_batch_internal(
...
+            nfts_metadata,
+        },
+    );
+
+    let mut ids = vec![];
+    for ids_subset in nft_ids.iter() {
+        for id in ids_subset.iter() {
+            ids.push(*id);
+        }
+    }
```
Comment (shamilsan) [inline tests/utils.rs]: Can we precalculate vec capacity? And don't use `iter` explicitly if it is not required.
```suggestion
    let mut ids = Vec::with_capacity(nft_ids.len() * ids_subset.len());
    for ids_subset in &nft_ids {
        for id in &ids_subset {
            ids.push(*id);
        }
    }
```

File: io/src/lib.rs [hunk 5]
```diff
@@ -7,54 +7,170 @@ use scale_info::TypeInfo;
...
+        /// Token's ID to burn.
+        id: TokenId,
+        /// Amount of burnt token.
+        amount: u128,
+        /// To which accounts to mint NFTs.
+        accounts: Vec<ActorId>,
+        /// NFTs' IDs.
+        nft_ids: Vec<Vec<TokenId>>,
+        /// NFTs metadata.
+        nfts_metadata: Vec<Vec<Option<TokenMetadata>>>,
```
Comment (shamilsan) [inline io/src/lib.rs]: Why to not combine these three vecs in one vec of structs?

**Review by shamilsan — COMMENTED** [review]
Please add to `CHANGELOG.md` info about changes.

**Comment by shamilsan** [comment]
Also, need to update `io` version too:

https://github.com/gear-dapps/multitoken/blob/master/io/Cargo.toml#L3

File: io/src/lib.rs (hunk 5 above)
Comment (Akado2009) [reply io/src/lib.rs]: Yeah, that's a good call!

File: src/lib.rs (hunk 3 above)
Comment (Akado2009) [reply src/lib.rs]: Yeah, should be metadata.len() for amount.

File: src/lib.rs [hunk 6]
```diff
@@ -106,11 +119,97 @@ impl SimpleMTKCore for SimpleMTK {
...
+        //         panic!("MTK: Mint to zero address")
+        //     }
+        //     if nft_ids[idx].len() != amount {
+        //         panic!("MTK: ids and amounts length mismatch")
+        //     }
+        //     metadata.iter().cloned().enumerate().for_each(|(i, meta)| {
+        //         self.mint_impl(&accounts[idx], &nft_ids[idx][i], NFT_COUNT, meta)
+        //     });
+        //     ids.append(&mut nft_ids[idx]);
+        // }
```
Comment (shamilsan) [inline src/lib.rs]: Dead code?

File: tests/mtk_tests.rs [hunk 7]
```diff
@@ -186,3 +191,48 @@ fn transfer_from_batch() {
...
+    //         media: Some(String::from("www.example.com/erc1155/kitty.png")),
+    //         reference: Some(String::from("www.example.com/erc1155/kitty")),
+    //     }),
+    //     Some(TokenMetadata {
+    //         title: Some(String::from("Kitty")),
+    //         description: Some(String::from("Just a test kitty #2")),
+    //         media: Some(String::from("www.example.com/erc1155/kitty.png")),
+    //         reference: Some(String::from("www.example.com/erc1155/kitty")),
+    //     }),
+    // ]];
```
Comment (shamilsan) [inline tests/mtk_tests.rs]: Dead code?

File: tests/utils.rs [hunk 8]
```diff
@@ -239,6 +239,39 @@ pub fn transfer_batch_internal(
...
+    from: u64,
+    token_id: u128,
+    amount: u128,
+    nfts: Vec<BurnToNFT>,
+) {
+    let mut ids = vec![];
+    for burn_info in &nfts {
+        for id in burn_info.nfts_ids.clone() {
+            ids.push(id);
+        }
```
Comment (shamilsan) [inline tests/utils.rs]: No need to clone the vector of `Copy`-type-elements that yields to `nfts.len()` useless allocations.

```suggestion
        for id in &burn_info.nfts_ids {
            ids.push(*id);
        }
```

File: tests/utils.rs [hunk 9]
```diff
@@ -239,6 +239,39 @@ pub fn transfer_batch_internal(
...
+            amount,
+            nfts,
+        },
+    );
+
+    let codec = MTKEvent::Transfer {
+        operator: from.into(),
+        from: ActorId::zero(),
+        to: ActorId::zero(),
+        ids: ids.to_vec(),
```
Comment (shamilsan) [inline tests/utils.rs]: It is already `Vec`.
```suggestion
        ids,
```

File: src/lib.rs [hunk 10]
```diff
@@ -106,11 +119,97 @@ impl SimpleMTKCore for SimpleMTK {
...
+                .nfts_metadata
+                .iter()
+                .cloned()
+                .enumerate()
+                .for_each(|(i, meta)| {
+                    self.mint_impl(&burn_info.account, &burn_info.nfts_ids[i], NFT_COUNT, meta)
+                });
+            for id in &burn_info.nfts_ids {
+                ids.push(*id);
+            }
```
Comment (shamilsan) [inline src/lib.rs]: If you don't use `nfts` later, you can consume it. This eliminates useless cloning and dereferencing.
```suggestion
        for burn_info in nfts {
            if burn_info.account.is_zero() {
                panic!("MTK: Mint to zero address");
            }
            if burn_info.nfts_ids.len() != burn_info.nfts_metadata.len() {
                panic!("MTK: ids and amounts length mismatch");
            }
            burn_info
                .nfts_metadata
                .into_iter()
                .enumerate()
                .for_each(|(i, meta)| {
                    self.mint_impl(&burn_info.account, &burn_info.nfts_ids[i], NFT_COUNT, meta)
                });
            for id in burn_info.nfts_ids {
                ids.push(id);
            }
```

**Review by shamilsan — COMMENTED** [review]
I don't see updates in `CHANGELOG.md`.

**Comment by Akado2009** [comment]
> I don't see updates in `CHANGELOG.md`.

Everything is fixed and changelog is introduced.

File: CHANGELOG.md [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+### Added
+- Transform function to convert FT tokens to multiple NFTs.
+### Changed.
+- Versions bumped.
+### Removed
+- JS wrappers.
+
+[Unreleased]: https://github.com/gear-dapps/multitoken/compare/0.1.0...HEAD
+[0.1.0]: https://github.com/gear-dapps/multitoken/compare/0.1.0...HEAD
+[0.2.0]: https://github.com/gear-dapps/multitoken/compare/0.2.0...HEAD
```
Comment (shamilsan) [inline CHANGELOG.md]: Pay attention to https://keepachangelog.com
```suggestion
## [0.3.0] - 2022-09-28
### Added
- Transform function to convert FT tokens to multiple NFTs.
### Changed.
- Versions bumped.
### Removed
- JS wrappers.

## [0.2.0] - 2022-09-04
### Added
- meta_state testing.
- JS wrappers.
### Changed
- All main dependencies' versions bumped and fixed.

## [0.1.0] - 2022-06-24
### Added
- Initial release.

[Unreleased]: https://github.com/gear-dapps/multitoken/compare/0.3.0...HEAD
[0.3.0]: https://github.com/gear-dapps/multitoken/compare/0.2.0...0.3.0
[0.2.0]: https://github.com/gear-dapps/multitoken/compare/0.1.0...0.2.0
[0.1.0]: https://github.com/gear-dapps/multitoken/compare/0803a5e...0.1.0
```

**Review by shamilsan — APPROVED** [review]

**Review by fluiderson — APPROVED** [review]

**Pull request merged**